File: pi/lib/herdr_workflow/findings.py
"""Pure finding schema validation and cross-verifier consolidation."""
import hashlib
import re


def validate_report_events(events, path):
    if len(events) > 31:
        raise SystemExit(f"report exceeds 30 findings plus verdict: {path}")
    for event in events:
        if event.get("type") == "verdict":
            continue
        if event.get("type") != "finding":
            raise SystemExit(f"report contains unsupported record type: {path}")
        if event.get("severity") not in {"critical", "warning", "info"} or not isinstance(event.get("path"), str) or not isinstance(event.get("line"), int) or not isinstance(event.get("detail"), str):
            raise SystemExit(f"invalid finding schema: {path}")
        for field, limit in (("detail", 1000), ("evidence", 2000), ("fix", 1000)):
            if field in event and (not isinstance(event[field], str) or len(event[field]) > limit):
                raise SystemExit(f"finding {field} exceeds {limit} characters: {path}")


def consolidate(events_by_role, prior_round, accepted_ids):
    """Dedupe findings across verifier roles, assign new/unfixed/fixed/accepted status.

    events_by_role: {role: [event, ...]} already-loaded JSONL events per role.
    prior_round: [finding, ...] from the previous round's history (may be empty).
    accepted_ids: set of finding ids marked accepted by a developer.
    Returns the flattened findings list (new + prior fixed/unfixed/accepted).
    """
    unique = {}
    for role, events in events_by_role.items():
        for event in events:
            if event.get("type") != "finding":
                continue
            detail = str(event.get("detail", ""))
            path = event.get("path")
            line = event.get("line")
            key = re.sub(r"\s+", " ", f"{path}:{line}:{detail}".lower())
            finding_id = str(event.get("id") or hashlib.sha256(key.encode()).hexdigest()[:12])
            unique.setdefault(finding_id, {"id": finding_id, "severity": event["severity"], "role": role, "detail": detail, "path": path, "line": line, "evidence": event.get("evidence"), "fix": event.get("fix")})
    previous = {item["id"] for item in prior_round if item.get("status") in {"new", "unfixed"}}
    findings = []
    for finding in unique.values():
        finding["status"] = "accepted" if finding["id"] in accepted_ids else "unfixed" if finding["id"] in previous else "new"
        findings.append(finding)
    for prior in prior_round:
        if prior["id"] not in unique and prior.get("status") in {"new", "unfixed"}:
            findings.append({**prior, "status": "fixed"})
    return findings

File: agentic-coding/src/workflow/findings.ts
// Pure finding schema validation and cross-verifier consolidation.
import { createHash } from 'node:crypto';

export const REPORT_CONTRACT = 'JSONL report contract: write 0-30 {"type":"finding","severity":"critical|warning|info","path":"<file>","line":<integer>,"detail":"<issue>","evidence":"<optional evidence>","fix":"<optional fix>"} records, then exactly one final {"type":"verdict","verdict":"PASS|FAIL"} record. Supported record types are finding and verdict only.';

export interface FindingEvent {
  type?: string;
  severity?: string;
  path?: string;
  line?: number;
  detail?: string;
  evidence?: string;
  fix?: string;
  id?: string;
  verdict?: string;
  [key: string]: unknown;
}

export interface Finding {
  id: string;
  severity?: string;
  role: string;
  detail: string;
  path?: string;
  line?: number;
  evidence?: string;
  fix?: string;
  status: string;
  [key: string]: unknown;
}

export function validateReportEvents(events: FindingEvent[], path: string): void {
  if (events.length > 31) throw new Error(`report exceeds 30 findings plus verdict: ${path}`);
  const verdicts = events.filter(event => event.type === 'verdict');
  if (verdicts.length !== 1 || events.at(-1)?.type !== 'verdict' || !['PASS', 'FAIL'].includes(verdicts[0]?.verdict ?? '')) throw new Error(`report must end with exactly one JSONL verdict PASS or FAIL: ${path}`);
  for (const event of events) {
    if (event.type === 'verdict') continue;
    if (event.type !== 'finding') throw new Error(`report contains unsupported record type: ${path}`);
    if (
      !['critical', 'warning', 'info'].includes(event.severity ?? '') ||
      typeof event.path !== 'string' ||
      !Number.isInteger(event.line) ||
      typeof event.detail !== 'string'
    ) {
      throw new Error(`invalid finding schema: ${path}`);
    }
    const limits: Array<[string, number]> = [
      ['detail', 1000],
      ['evidence', 2000],
      ['fix', 1000],
    ];
    for (const [field, limit] of limits) {
      if (field in event) {
        const value = (event as Record<string, unknown>)[field];
        if (typeof value !== 'string' || value.length > limit) {
          throw new Error(`finding ${field} exceeds ${limit} characters: ${path}`);
        }
      }
    }
  }
}

/**
 * Dedupe findings across verifier roles, assign new/unfixed/fixed/accepted status.
 *
 * eventsByRole: {role: [event, ...]} already-loaded JSONL events per role.
 * priorRound: [finding, ...] from the previous round's history (may be empty).
 * acceptedIds: set of finding ids marked accepted by a developer.
 * Returns the flattened findings list (new + prior fixed/unfixed/accepted).
 */
export function consolidate(eventsByRole: Record<string, FindingEvent[]>, priorRound: Finding[], acceptedIds: Set<string>): Finding[] {
  const unique = new Map<string, Finding>();
  for (const [role, events] of Object.entries(eventsByRole)) {
    for (const event of events) {
      if (event.type !== 'finding') continue;
      const detail = String(event.detail ?? '');
      const path = event.path;
      const line = event.line;
      const key = `${path}:${line}:${detail}`.toLowerCase().replace(/\s+/g, ' ');
      const findingId = String(event.id || createHash('sha256').update(key).digest('hex').slice(0, 12));
      if (!unique.has(findingId)) {
        unique.set(findingId, {
          id: findingId,
          severity: event.severity,
          role,
          detail,
          path,
          line,
          evidence: event.evidence,
          fix: event.fix,
          status: 'new',
        });
      }
    }
  }
  const previous = new Set(priorRound.filter(item => item.status === 'new' || item.status === 'unfixed').map(item => item.id));
  const findings: Finding[] = [];
  for (const finding of unique.values()) {
    finding.status = acceptedIds.has(finding.id) ? 'accepted' : previous.has(finding.id) ? 'unfixed' : 'new';
    findings.push(finding);
  }
  for (const prior of priorRound) {
    if (!unique.has(prior.id) && (prior.status === 'new' || prior.status === 'unfixed')) {
      findings.push({ ...prior, status: 'fixed' });
    }
  }
  return findings;
}
